README.txt
This package provides a class to control the GPIO on a Raspberry Pi, ported
to work on the UP board hardware.

Note that this module is unsuitable for real-time or timing critical applications.  This is because you
can not predict when Python will be busy garbage collecting.  It also runs under the Linux kernel which
is not suitable for real time applications - it is multitasking O/S and another process may be given
priority over the CPU, causing jitter in your program.  If you are after true real-time performance and
predictability, buy yourself an Arduino http://www.arduino.cc !

Note that the current release does not support SPI, I2C, hardware PWM or serial functionality on the RPi yet.
This is planned for the near future - watch this space!  One-wire functionality is also planned.

Although hardware PWM is not available yet, software PWM is available to use on all channels.

For examples and documentation, visit http://sourceforge.net/p/raspberry-gpio-python/wiki/Home/



CHANGELOG.txt
Change Log
==========

0.6.3+ubi1
----------
- Ported to the UP Squared board

0.6.3
-------
- Fix code so it builds under PyPy (Gasper Zejn)
- os.system breaks event detection - Matt Kimball (issue 127)

0.6.2
-----
- Rewrote Debian packaging mechanism
- RPI_INFO reports Pi 3
- Changed module layout - moved C components to RPi._GPIO

0.6.1
-----
- Update RPI_INFO to detect more board types
- Issue 118 - add_event_detect sometimes gives runtime error with unpriv user 
- Issue 120 - setmode() remembers invalid mode

0.6.0a3
-------
- Now uses /dev/gpiomem if available to avoid being run as root
- Fix warnings with pull up/down on pins 3/5
- Correct base address on Pi 2 when devicetree is disabled
- caddr_t error on compile (Issue 109)
- Error on invalid parameters to setup() (issue 93)
- Add timeout parameter to wait_for_edge() (issue 91)

0.5.11
------
- Fix - pins > 26 missing when using BOARD mode
- Add getmode()
- Raise exception when a mix of modes is used
- GPIO.cleanaup() unsets the current pin mode

0.5.10
------
- Issue 95 - support RPi 2 boards
- Introduce RPI_INFO
- Deprecate RPI_REVISION
- Issue 97 - fixed docstring for setup()

0.5.9
-----
- Issue 87 - warn about pull up/down on i2c pins
- Issue 86/75 - wait_for_edge() bugfix
- Issue 84 - recognise RPi properly when using a custom kernel
- Issue 90 - cleanup() on a list/tuple of channels

0.5.8
-----
- Allow lists/tuples of channels in GPIO.setup()
- GPIO.output() now allows lists/tuples of values
- GPIO.wait_for_edge() bug fixes (issue 78)

0.5.7
-----
- Issue 67 - speed up repeated calls to GPIO.wait_for_event()
- Added bouncetime keyword to GPIO.wait_for_event()
- Added extra edge/interrupt unit tests
- GPIO.wait_for_event() can now be mixed with GPIO.add_event_detect()
- Improved cleanups of events
- Issue 69 resolved

0.5.6
-----
- Issue 68 - support for RPi Model B+
- Fix gpio_function()

0.5.5
-----
- Issue 52 - 'unallocate' a channel
- Issue 35 - use switchbounce with GPIO.event_detected()
- Refactored events code
- Rewrote tests to use unittest mechanism and new test board with loopbacks
- Fixed adding events after a GPIO.cleanup()
- Issue 64 - misleading /dev/mem permissions error
- Issue 59 - name collision with PWM constant and class

0.5.4
-----
- Changed release status (from alpha to full release)
- Warn when GPIO.cleanup() used with nothing to clean up (issue 44)
- Avoid collisions in constants (e.g. HIGH / RISING / PUD_DOWN)
- Accept BOARD numbers in gpio_function (issue 34)
- More return values for gpio_function (INPUT, OUTPUT, SPI, I2C, PWM, SERIAL, UNKNOWN)
- Tidy up docstrings
- Fix /dev/mem access error with gpio_function

0.5.3a
------
- Allow pydoc for non-root users (issue 27)
- Fix add_event_detect error when run as daemon (issue 32)
- Simplified exception types
- Changed from distribute to pip

0.5.2a
------
- Added software PWM (experimental)
- Added switch bounce handling to event callbacks
- Added channel number parameter to event callbacks (issue 31)
- Internal refactoring and code tidy up

0.5.1a
------
- Fixed callbacks for multiple GPIOs (issue 28)

0.5.0a
------
- Added new edge detection events (interrupt handling)
  - Added add_event_detect()
  - Added remove_event_detect()
  - Added add_event_callback()
  - Added wait_for_edge()
- Removed old experimental event functions
  - Removed set_rising_event()
  - Removed set_falling_event()
  - Removed set_high_event()
  - Removed set_low_event()
- Changed event_detected() for new edge detection functionality
- input() now returns 0/LOW == False or 1/HIGH == True (integers) instead of False or True (booleans).
- Fix error on repeated import (issue 3)
- Change SetupException to a RuntimeError so it can be caught on import (issue 25, Chris Hager <chris@linuxuser.at>)
- Improved docstrings of functions

0.4.2a
------
- Fix for installing on Arch Linux (Python 3.3) (issue 20)
- Initial value when setting a channel as an output (issue 19)

0.4.1a
------
- Added VERSION
- Permit input() of channels set as outputs (Eric Ptak <trouch@trouch.com>)

0.4.0a
------
- Added support for Revision 2 boards
- Added RPI_REVISION
- Added cleanup() function and removed automatic reset functionality on program exit
- Added get_function() to read existing GPIO channel functionality (suggestion from Eric Ptak <trouch@trouch.com>)
- Added set_rising_event()
- Added set_falling_event()
- Added set_high_event()
- Added set_low_event()
- Added event_detected()
- Added test/test.py
- Converted debian to armhf
- Fixed C function short_wait() (thanks to Thibault Porteboeuf <thibaultporteboeuf@gmail.com>)

0.3.1a
------
- Fixed critical bug with swapped high/low state on outputs
- Added pull-up / pull-down setup functionality for inputs

0.3.0a
------
- Rewritten as a C extension
- Now uses /dev/mem and SoC registers instead of /sys/class/gpio
- Faster!
- Make call to GPIO.setmode() mandatory
- Added GPIO.HIGH and GPIO.LOW constants

0.2.0
-----
- Changed status from alpha to beta
- Added setmode() to be able to use BCM GPIO 00.nn channel numbers
- Renamed InvalidPinException to InvalidChannelException

0.1.0
------
- Fixed direction bug
- Added MANIFEST.in (to include missing file)
- Changed GPIO channel number to pin number
- Tested and working!

0.0.3a
------
- Added GPIO table
- Refactored
- Fixed a few critical bugs
- Still completely untested!

0.0.2a
------
- Internal refactoring.  Still completely untested!

0.0.1a
------
- First version.  Completely untested until I can get hold of a Raspberry Pi!



source/c_gpio.h
/*
Copyright (c) 2012-2015 Ben Croston

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

int setup(void);
void setup_gpio(int gpio, int direction, int pud);
int gpio_function(int gpio);
void output_gpio(int gpio, int value);
int input_gpio(int gpio);
void set_rising_event(int gpio, int enable);
void set_falling_event(int gpio, int enable);
void set_high_event(int gpio, int enable);
void set_low_event(int gpio, int enable);
int eventdetected(int gpio);
void cleanup(void);

#define SETUP_OK           0
#define SETUP_DEVMEM_FAIL  1
#define SETUP_MALLOC_FAIL  2
#define SETUP_MMAP_FAIL    3
#define SETUP_CPUINFO_FAIL 4
#define SETUP_NOT_RPI_FAIL 5
#define SETUP_EXPORT_FAIL  6

#define INPUT  1 // is really 0 for control register!
#define OUTPUT 0 // is really 1 for control register!
#define ALT0   4

#define HIGH 1
#define LOW  0

#define PUD_OFF  0
#define PUD_DOWN 1
#define PUD_UP   2


source/cpuinfo.c
/*
Copyright (c) 2012-2016 Ben Croston

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "cpuinfo.h"

int get_rpi_info(rpi_info *info)
{
   // TODO - parse DMI data to obtain correct board information
   info->ram = "Unknown";
   info->manufacturer = "Unknown";
   info->processor = "Unknown";
   info->type = "Unknown";
   strcpy(info->revision, "0000");
   // assume RPi2 compatibility
   info->p1_revision = 3;
   return 0;
}


create_gpio_user_permissions.py
import grp
import subprocess

def ensure_gpiogroup():
    try:
        grp.getgrnam('gpio')
    except KeyError:
        print('GPIO group does not exist - creating...')
        subprocess.call(['groupadd', '-f', '-r', 'gpio'])
        subprocess.call(['adduser', 'ubilinux', 'gpio'])
        # in future, also for groups:
        #   spi
        #   i2c
        add_udev_rules()

def add_udev_rules():
    with open('/etc/udev/rules.d/99-gpio.rules','w') as f:
        f.write("""SUBSYSTEM=="bcm2835-gpiomem", KERNEL=="gpiomem", GROUP="gpio", MODE="0660"
SUBSYSTEM=="gpio", KERNEL=="gpiochip*", ACTION=="add", PROGRAM="/bin/sh -c 'chown root:gpio /sys/class/gpio/export /sys/class/gpio/unexport ; chmod 220 /sys/class/gpio/export /sys/class/gpio/unexport'"
SUBSYSTEM=="gpio", KERNEL=="gpio*", ACTION=="add", PROGRAM="/bin/sh -c 'chown root:gpio /sys%p/active_low /sys%p/direction /sys%p/edge /sys%p/value ; chmod 660 /sys%p/active_low /sys%p/direction /sys%p/edge /sys%p/value'"
""")

if __name__ == '__main__':
    ensure_gpiogroup()


source/c_gpio.c
/*
Copyright (c) 2012-2015 Ben Croston

Adapted for use on UP board (http://www.up-board.org/)
Copyright (c) 2015 Emutex Ltd
Author: Dan O'Donovan <dan@emutex.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <string.h>
#include <time.h>
#include <sys/time.h>
#include "c_gpio.h"

struct gpio_chip_mmr {
    uint64_t phys;
    void *virt;
    unsigned size;
    int fd;
};
	
struct up_gpio_pin_desc {
    struct gpio_chip_mmr *mmr;
    unsigned pad;
    unsigned offset;
    int val_fd;
    int dir_fd;
    int initialised;
};

static struct gpio_chip_mmr cht_gpio_mmr_SW = {
    .phys = 0xfed80000,
    .size = 0x8000
};
static struct gpio_chip_mmr cht_gpio_mmr_N  = {
    .phys = 0xfed88000,
    .size = 0x8000
};
static struct gpio_chip_mmr cht_gpio_mmr_E  = {
    .phys = 0xfed90000,
    .size = 0x8000
};
static struct gpio_chip_mmr cht_gpio_mmr_SE = {
    .phys = 0xfed98000,
    .size = 0x8000
};

#define CHT_INTSTAT_OFFSET 0x300

#define CHT_PADREG0_INTSEL_SHIFT   28
#define CHT_PADREG0_INTSEL_MASK    (0xf << CHT_PADREG0_INTSEL_SHIFT)

#define CHT_PADREG1_INVRXTX_MASK   0xf0
#define CHT_PADREG1_INVRXTX_RXDATA 0x40
#define CHT_PADREG1_INTCFG_MASK    7
#define CHT_PADREG1_INTCFG_FALLING 1
#define CHT_PADREG1_INTCFG_RISING  2
#define CHT_PADREG1_INTCFG_BOTH    3
#define CHT_PADREG1_INTCFG_LEVEL   4

#define CHT_PADREG_OFFSET(pad) \
    (0x4400 + (0x400 * ((pad) / 15)) + (0x8 * ((pad) % 15)))

#define CHT_PAD(m, p)               \
{                                   \
    .mmr = (m),                     \
    .pad = (p),                     \
    .offset = CHT_PADREG_OFFSET(p), \
}

#define CHT_PADREG_0_GPIO_TX_MASK 0x2
#define CHT_PADREG_0_GPIO_RX_MASK 0x1

#define CHT_PADREG(base, offset) \
    *((volatile uint32_t *)(base + offset))

static struct up_gpio_pin_desc up_gpio_pins[] = {
	CHT_PAD(&cht_gpio_mmr_SW, 61), /*  0 */
	CHT_PAD(&cht_gpio_mmr_SW, 65), /*  1 */
	CHT_PAD(&cht_gpio_mmr_SW, 60), /*  2 */
	CHT_PAD(&cht_gpio_mmr_SW, 63), /*  3 */
	CHT_PAD(&cht_gpio_mmr_E,  21), /*  4 */
	CHT_PAD(&cht_gpio_mmr_E,  24), /*  5 */
	CHT_PAD(&cht_gpio_mmr_E,  15), /*  6 */
	CHT_PAD(&cht_gpio_mmr_SE, 79), /*  7 */
	CHT_PAD(&cht_gpio_mmr_SE,  7), /*  8 */
	CHT_PAD(&cht_gpio_mmr_SE,  3), /*  9 */
	CHT_PAD(&cht_gpio_mmr_SE,  6), /* 10 */
	CHT_PAD(&cht_gpio_mmr_SE,  4), /* 11 */
	CHT_PAD(&cht_gpio_mmr_SE,  5), /* 12 */
	CHT_PAD(&cht_gpio_mmr_SE,  1), /* 13 */
	CHT_PAD(&cht_gpio_mmr_SW, 20), /* 14 */
	CHT_PAD(&cht_gpio_mmr_SW, 16), /* 15 */
	CHT_PAD(&cht_gpio_mmr_N,   6), /* 16 */
	CHT_PAD(&cht_gpio_mmr_E,  18), /* 17 */
	CHT_PAD(&cht_gpio_mmr_SW, 31), /* 18 */
	CHT_PAD(&cht_gpio_mmr_SW, 35), /* 19 */
	CHT_PAD(&cht_gpio_mmr_SW, 33), /* 20 */
	CHT_PAD(&cht_gpio_mmr_SW, 30), /* 21 */
	CHT_PAD(&cht_gpio_mmr_N,   3), /* 22 */
	CHT_PAD(&cht_gpio_mmr_N,   2), /* 23 */
	CHT_PAD(&cht_gpio_mmr_N,   1), /* 24 */
	CHT_PAD(&cht_gpio_mmr_SW, 21), /* 25 */
	CHT_PAD(&cht_gpio_mmr_N,   7), /* 26 */
	CHT_PAD(&cht_gpio_mmr_SW, 17), /* 27 */
};	  

static int export_fd = -1;

/************* /sys/class/gpio functions ************/

int setup(void)
{
    if ((export_fd = open("/sys/class/gpio/export", O_WRONLY)) < 0)
       return SETUP_EXPORT_FAIL;

    return SETUP_OK;
}

int init_gpio(int gpio)
{
    int len;
    char str_gpio[3];
    char filename[33];
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    snprintf(filename, sizeof(filename),
             "/sys/class/gpio/gpio%d/value", gpio);
    if (0 != access(filename, F_OK))
    {
        len = snprintf(str_gpio, sizeof(str_gpio), "%d", gpio);
        lseek(export_fd, 0, SEEK_SET);
        if (write(export_fd, str_gpio, len) != len)
            return SETUP_EXPORT_FAIL;
    }

    if (!(pd->mmr->virt))
    {
        if ((pd->mmr->fd = open("/dev/mem", O_RDWR | O_SYNC)) >= 0)
        {
            void *virt = mmap(NULL, pd->mmr->size, PROT_READ | PROT_WRITE,
                              MAP_SHARED, pd->mmr->fd, pd->mmr->phys);
            if (MAP_FAILED != virt)
                pd->mmr->virt = virt;
        }
    }

    // arbitary delay to allow udev time to set user permissions
    struct timespec delay;
    delay.tv_sec = 0;
    delay.tv_nsec = 50000000L; // 50ms
    nanosleep(&delay, NULL);

    snprintf(filename, sizeof(filename),
             "/sys/class/gpio/gpio%d/direction", gpio);
    if ((pd->dir_fd = open(filename, O_RDWR | O_SYNC)) < 0)
        return SETUP_EXPORT_FAIL;

    // Fall back to sysfs access if unable to access /dev/mem
    if (!(pd->mmr->virt))
    {
        snprintf(filename, sizeof(filename),
                 "/sys/class/gpio/gpio%d/value", gpio);
        if ((pd->val_fd = open(filename, O_RDWR | O_SYNC)) < 0)
            return SETUP_EXPORT_FAIL;
    }

    pd->initialised = 1;

    return SETUP_OK;
}

void clear_event_detect(int gpio)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    if (pd->mmr->virt)
    {
        unsigned intr_line = CHT_PADREG(pd->mmr->virt, pd->offset);
        intr_line &= CHT_PADREG0_INTSEL_MASK;
        intr_line >>= CHT_PADREG0_INTSEL_SHIFT;
        
        CHT_PADREG(pd->mmr->virt, CHT_INTSTAT_OFFSET) = (1 << intr_line);
    }
}

int eventdetected(int gpio)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    if (pd->mmr->virt)
    {
        unsigned intr_line = CHT_PADREG(pd->mmr->virt, pd->offset);
        intr_line &= CHT_PADREG0_INTSEL_MASK;
        intr_line >>= CHT_PADREG0_INTSEL_SHIFT;
        
        return !!(CHT_PADREG(pd->mmr->virt, CHT_INTSTAT_OFFSET)
                  & (1 << intr_line));
    }
    return 0;
}

void set_rising_event(int gpio, int enable)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    if (pd->mmr->virt)
    {
        uint32_t value = CHT_PADREG(pd->mmr->virt, pd->offset + 0x4);
        value &= ~(CHT_PADREG1_INTCFG_MASK | CHT_PADREG1_INVRXTX_MASK);
        value |= CHT_PADREG1_INTCFG_RISING;
        CHT_PADREG(pd->mmr->virt, pd->offset + 0x4) = value;
    }
}

void set_falling_event(int gpio, int enable)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    if (pd->mmr->virt)
    {
        uint32_t value = CHT_PADREG(pd->mmr->virt, pd->offset + 0x4);
        value &= ~(CHT_PADREG1_INTCFG_MASK | CHT_PADREG1_INVRXTX_MASK);
        value |= CHT_PADREG1_INTCFG_FALLING;
        CHT_PADREG(pd->mmr->virt, pd->offset + 0x4) = value;
    }
}

void set_high_event(int gpio, int enable)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    if (pd->mmr->virt)
    {
        uint32_t value = CHT_PADREG(pd->mmr->virt, pd->offset + 0x4);
        value &= ~(CHT_PADREG1_INTCFG_MASK | CHT_PADREG1_INVRXTX_MASK);
        value |= CHT_PADREG1_INTCFG_LEVEL;
        CHT_PADREG(pd->mmr->virt, pd->offset + 0x4) = value;
    }
}

void set_low_event(int gpio, int enable)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    if (pd->mmr->virt)
    {
        uint32_t value = CHT_PADREG(pd->mmr->virt, pd->offset + 0x4);
        value &= ~(CHT_PADREG1_INTCFG_MASK | CHT_PADREG1_INVRXTX_MASK);
        value |= (CHT_PADREG1_INTCFG_LEVEL | CHT_PADREG1_INVRXTX_RXDATA);
        CHT_PADREG(pd->mmr->virt, pd->offset + 0x4) = value;
    }
}

void set_pullupdn(int gpio, int pud)
{
    // TODO (current board design has fixed external pull-ups)
}

void setup_gpio(int gpio, int direction, int pud)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    set_pullupdn(gpio, pud);

    if (pd->dir_fd >= 0)
    {
        lseek(pd->dir_fd, 0, SEEK_SET);
        if (direction == INPUT)
            write(pd->dir_fd, "in", 2);
        else
            write(pd->dir_fd, "out", 3);
    }
}

int gpio_function(int gpio)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    if (pd->dir_fd >= 0)
    {
        char str_dir[3];

        lseek(pd->dir_fd, 0, SEEK_SET);
        if (read(pd->dir_fd, str_dir, sizeof(str_dir)) > 0)
            return (str_dir[0] == 'i') ? 0 : 1;
    }

    return 0;
}

void output_gpio(int gpio, int value)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    if (pd->mmr->virt)
    {
        if (value)
            CHT_PADREG(pd->mmr->virt, pd->offset) |= CHT_PADREG_0_GPIO_TX_MASK;
        else
            CHT_PADREG(pd->mmr->virt, pd->offset) &= ~CHT_PADREG_0_GPIO_TX_MASK;
    }
    else if (pd->val_fd >= 0)
    {
        lseek(pd->val_fd, 0, SEEK_SET);
        write(pd->val_fd, value ? "1" : "0", 1);
    }
}

int input_gpio(int gpio)
{
    struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

    if (!pd->initialised)
        init_gpio(gpio);

    if (pd->mmr->virt)
        return !!(CHT_PADREG(pd->mmr->virt, pd->offset)
                  & CHT_PADREG_0_GPIO_RX_MASK);
    else if (pd->val_fd >= 0)
    {
        char str_val[2];

        lseek(pd->val_fd, 0, SEEK_SET);
        if (read(pd->val_fd, str_val, sizeof(str_val)) > 0)
            return (str_val[0] == '1');
    }

    return 0;
}

void cleanup(void)
{
    unsigned int gpio;
    int unexport_fd;

    close(export_fd);
    export_fd = -1;

    if ((unexport_fd = open("/sys/class/gpio/unexport", O_WRONLY)) < 0)
       return;

    for (gpio = 0; gpio < sizeof(up_gpio_pins)/sizeof(up_gpio_pins[0]); gpio++)
    {
        int len;
        char str_gpio[3];
        struct up_gpio_pin_desc *pd = &up_gpio_pins[gpio];

        if (pd->initialised)
        {
            pd->initialised = 0;

            if (pd->mmr->virt)
            {
                munmap(pd->mmr->virt, pd->mmr->size);
                close(pd->mmr->fd);
                pd->mmr->virt = NULL;
            }

            if (pd->dir_fd >= 0)
                close(pd->dir_fd);

            if (pd->val_fd >= 0)
                close(pd->val_fd);

            len = snprintf(str_gpio, sizeof(str_gpio), "%d", gpio);
            lseek(unexport_fd, 0, SEEK_SET);
            if (write(unexport_fd, str_gpio, len) != len)
            {
                close(unexport_fd);
                return;
            }
        }
    }

    close(unexport_fd);
}
